### ISSUE TITLE
Wrong modulus being used in Mod

### ISSUE THREAD

**Issue by gilgamec** [Bodigrim/arithmoi#163]
I'm getting a problem where modular operations are being performed with the wrong modulus. An example:
```haskell
{-# LANGUAGE PolyKinds, RankNTypes #-}

module Main where

import Math.NumberTheory.Moduli
import GHC.TypeNats hiding (Mod(..))

modl :: (forall m. KnownNat m => Mod m) -> Integer -> Integer
modl m n = case someNatVal (fromInteger n) of
  SomeNat p -> getVal (m `asProxyOf` p)

asProxyOf :: f a -> proxy a -> f a
asProxyOf = const

main :: IO ()
main = print $ (fromInteger 127 ^ 31336) `modl` 31337

otherTests :: [Integer]
otherTests =
  [
    (fromInteger 3161 ^ 2267) `modl` 963460043520452550503
  , (fromInteger 4 ^ 712) `modl` 22675946387565680267
  , (fromInteger 19128 ^ 583658) `modl` 782014287063416516807
  ]
```
(The numbers are arbitrary.)

Running from `ghci`, `main` produces the correct answer, 1. When compiled, the answer is `(127 ^ 31336)` modulo one of the primes in `otherTests`. If `otherTests` is commented out, we get the correct answer again. Interestingly, even if only two of the three tests in `otherTests` are commented out, the program still prints 1.

**Comment by Bodigrim** [comment]
I can reproduce the issue with GHC 8.6.3 and `arithmoi` HEAD.

```
$ ghc -fforce-recomp Mod.hs && ./Mod
1
$ ghc -fforce-recomp -O Mod.hs && ./Mod
6174800926042330116
```

This is very weird. Thanks for reporting, I'll investigate it after Easter holidays. 

In the meantime, please use `Math.NumberTheory.Powers.Modular.powMod` for modular powers.

**Comment by Bodigrim** [comment]
Seems to be GHC bug.
Filed as https://gitlab.haskell.org/ghc/ghc/issues/16586

**Comment by Bodigrim** [comment]
This has been fixed in GHC 8.8. Thanks for reporting.
